platform/android/net_socket_android.cpp

```diff
@@ -0,0 +1,105 @@
+/*  net_socket_android.cpp                                               */
+/*************************************************************************/
+/*                       This file is part of:                           */
+/*                           GODOT ENGINE                                */
+/*                      https://godotengine.org                          */
+/*************************************************************************/
+/* Copyright (c) 2007-2019 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2019 Godot Engine contributors (cf. AUTHORS.md)    */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "net_socket_android.h"
+
+#include "java_godot_io_wrapper.h"
+#include "os_android.h"
+
+NetSocket *NetSocketAndroid::_create_func() {
+	return memnew(NetSocketAndroid);
+}
+
+void NetSocketAndroid::make_default() {
+	_create = _create_func;
+}
+
+NetSocketAndroid::NetSocketAndroid() :
+		wants_broadcast(false),
+		multicast_groups(0) {
+}
+
+NetSocketAndroid::~NetSocketAndroid() {
+	close();
+}
+
+void NetSocketAndroid::close() {
+	NetSocketPosix::close();
+	if (wants_broadcast)
+		((OS_Android *)OS::get_singleton())->get_godot_io_java()->multicast_lock_release();
+	if (multicast_groups)
+		((OS_Android *)OS::get_singleton())->get_godot_io_java()->multicast_lock_release();
+	wants_broadcast = false;
+	multicast_groups = 0;
+}
+
+Error NetSocketAndroid::set_broadcasting_enabled(bool p_enabled) {
```

Comment:
The logic is very similar for all three methods. Seems like you can merge `wants_broadcast` and `multicast_group` into a single counter and define a private helper method that all three methods call to:

- increment the counter and acquire the multicast lock when joining a multicast group or enabling broadcast.

- decrement the counter and release the multicast lock when leaving a multicast group or disabling broadcast.